src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -1419,6 +1595,266 @@ DatabaseShardImp::removeFailedShard(std::shared_ptr<Shard> shard)
     setFileStats();
 }
 
+std::uint32_t
+DatabaseShardImp::getShardBoundaryIndex(std::lock_guard<std::mutex> const&)
+{
+    auto const validIndex = app_.getLedgerMaster().getValidLedgerIndex();
+
+    // Shards with an index earlier than recentShardBoundaryIndex
+    // are considered historical. The three shards at or later than
+    // this index consist of the two most recently validated shards
+    // and the shard still in the process of being built by live
+    // transactions.
+    return NodeStore::seqToShardIndex(validIndex, ledgersPerShard_) - 1;
+}
+
+std::uint32_t
+DatabaseShardImp::countHistoricalShards(std::lock_guard<std::mutex> const& lock)
+{
+    auto const recentShardBoundaryIndex = getShardBoundaryIndex(lock);
+
+    return std::count_if(
+        shards_.begin(),
+        shards_.end(),
+        [recentShardBoundaryIndex](decltype(shards_)::value_type const& entry) {
```

Comment:
Could use `auto const& entry` for the param type here and get rid of the `decltype`

Reply:
Done